Project import simple implementation

ActiveChooN opened the pull request:
<!---
Copyright (C) 2020-2021 Intel Corporation

SPDX-License-Identifier: MIT
-->

<!-- Raised an issue to propose your change (https://github.com/opencv/cvat/issues).
It will help avoiding duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [CONTRIBUTION](https://github.com/opencv/cvat/blob/develop/CONTRIBUTING.md)
guide. -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->

This PR provides the whole dataset (data and annotation) import to the project function. In this iteration splitting data to the task just made by subsets and will be improved in future releases. 

Resolves #2331, #1768

### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### TODO Checklist
- [x] Tracks support
- [x] All formats import support
- [x] Import status in UI

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable by a reason then ~~explicitly strikethrough~~ the whole
line. If you don't do that github will show incorrect process for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
~~- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly~~
- [x] I have added tests to cover my changes
~~- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))~~
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

zhiltsov-max commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -391,14 +427,25 @@ def tags(self):
             for attrib in _tag['attributes']
-            if self._get_attribute_id(label_id, attrib.name)]
+            if self._get_attribute_id(label_id, attrib.name) or (
+                self.soft_attribute_import and attrib.name not in {'occluded', 'outside', 'keyframe', 'track_id'}
```
Consider extracting a variable with these attributes.

ActiveChooN commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -1249,27 +1392,77 @@ def import_dm_annotations(dm_dataset, task_data):
+    else:
+        for label in dataset.categories()[datum_annotation.AnnotationType.label].items:
+            if not project_annotation.db_project.label_set.filter(name=label.name).exists():
+                raise CvatImportError(f'Target project does not have label with name "{label.name}"')
```
@zhiltsov-max, yes, it's planned to do

ActiveChooN commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -508,40 +565,44 @@ def match_frame_fuzzy(self, path):
+    Frame = NamedTuple('Frame',
+        [('idx', int), ('id', int), ('frame', int), ('name', str), ('width', int), ('height', int), ('labeled_shapes', List[Union[LabeledShape, TrackedShape]]), ('tags', List[Tag]), ('task_id', int), ('subset', str)],
+    )
+    Frame.__new__.__defaults__ = (None, None)
```
This is not needed since datumaro extractors are used for all formats, so I deleted it.

ActiveChooN commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -104,6 +107,39 @@ def _import_attribute(self, label_id, attribute):
+            else:
+                value = str(value)
+                if value.lower() in {'true', 'false'}:
+                    value = value.lower() == 'true'
```
It is, but this looks like a more general solution for

bsekachev commented on cvat-core/src/annotations.js:
```diff
@@ -284,6 +284,22 @@
     }
 
+    async function importDataset(instance, format, file, updateStatusCallback) {
+        if (!(format instanceof String || typeof format === 'string')) {
```
Do we need both checks here?

bsekachev commented on cvat-core/src/annotations.js:
```diff
@@ -284,6 +284,22 @@
+            throw new ArgumentError('Format must be a string');
+        }
+        if (!(instance instanceof Project)) {
+            throw new ArgumentError('Instance should ne a Project isntance');
```
```suggestion
            throw new ArgumentError('Instance should be a Project instance');
```

bsekachev commented on cvat-core/src/annotations.js:
```diff
@@ -284,6 +284,22 @@
+            throw new ArgumentError('Instance should ne a Project isntance');
+        }
+        if (!(typeof updateStatusCallback === 'function' || updateStatusCallback === null)) {
+            throw new ArgumentError('Callback should ne a function or null');
```
```suggestion
            throw new ArgumentError('Callback should be a function or null');
```

bsekachev commented on cvat-core/src/annotations.js:
```diff
@@ -284,6 +284,22 @@
+        if (!(instance instanceof Project)) {
+            throw new ArgumentError('Instance should ne a Project isntance');
+        }
+        if (!(typeof updateStatusCallback === 'function' || updateStatusCallback === null)) {
```
You can set a default argument as an empty function: ``() => void`` and left one check only

bsekachev commented on cvat-core/src/annotations.js:
```diff
@@ -284,6 +284,22 @@
+            throw new ArgumentError('Callback should ne a function or null');
+        }
+        if (!(file instanceof File && file.name.split('.').reverse()[0])) {
+            throw new ArgumentError('File should be file instance with ZIP extension');
```
Why don't we support other archive extensions?
I don't think it is a big deal

bsekachev replied:
Can we check file.type? (as in cvat-ui) to be equal one of: 'application/zip', 'application/x-zip-compressed'

bsekachev commented on cvat-core/src/annotations.js:
```diff
@@ -284,6 +284,22 @@
         return result;
     }
 
+    async function importDataset(instance, format, file, updateStatusCallback) {
```
Unnecesassary ``async``. The method does not use ``await``

bsekachev commented on cvat-core/src/server-proxy.js:
```diff
@@ -499,6 +499,47 @@
+                                proxy: config.proxy,
+                            });
+                            if (response.status === 202) {
+                                if (typeof callback === 'function') callback(response.data);
```
```suggestion
                                if (callback) callback(response.data);
```

we already know, that it is a function or null

bsekachev commented on cvat-core/src/server-proxy.js:
```diff
@@ -499,6 +499,47 @@
+                            }
+                        } catch (error) {
+                            if (error.response.status === 404) {
+                                try {
+                                    await Axios.post(`${url}?format=${format}`, formData, {
+                                        proxy: config.proxy,
+                                    });
+                                    setTimeout(request, 2000);
+                                } catch (_error) {
+                                    reject(generateError(error));
+                                }
```
I would suggest to put it before getting status for the first time. In this way we always do unnesassary request first.

bsekachev commented on cvat-ui/src/actions/import-actions.ts:
```diff
@@ -0,0 +1,59 @@
+    ),
+};
+
+export const importDatasetAsync = (instance: any, format: string, file: File): ThunkAction => async (dispatch) => {
+    try {
+        const store = getCVATStore();
+        const state: CombinedState = store.getState();
```
```suggestion
export const importDatasetAsync = (instance: any, format: string, file: File): ThunkAction => async (dispatch, getState) => {
    try {
        const state: CombinedState = getState();
```

bsekachev replied:
This way should work :)

bsekachev commented on cvat-ui/src/actions/import-actions.ts:
```diff
@@ -0,0 +1,59 @@
+            throw Error('Only one importing of dataset allowed at the same time');
+        }
+        dispatch(importActions.importDataset(instance, format));
+        await instance.annotations.importDataset(format, file, (response: any) => (
```
Instead of passing the whole response to this callback, I would suggest passing something more specific (progress, message) for example.

bsekachev commented on cvat-ui/src/actions/import-actions.ts:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    dispatch(importActions.importDatasetSuccess(instance, format));
+    dispatch(getProjectsAsync({ id: instance.id }));
```
If always need to update current project instance after importing a dataset, it would be logically if instance.annotations.importDataset returns updated project.

ActiveChooN replied:
I don't sure it would be so clear

bsekachev commented on cvat-ui/src/components/import-dataset-modal/import-dataset-status-modal.tsx:
```diff
@@ -0,0 +1,34 @@
+            visible={!!id}
+            closable={false}
+            footer={null}
+            className='cvat-modal-import-project-status'
```
Would suggest adding ``destroyOnClose`` flag

bsekachev replied:
```suggestion
            className='cvat-modal-import-dataset-status'
```

bsekachev commented on cvat-ui/src/components/import-dataset-modal/import-dataset-status-modal.tsx:
```diff
@@ -0,0 +1,34 @@
+
+    return (
+        <Modal
+            title={`Importing dataset for project #${id}`}
```
```suggestion
            title={`Importing a dataset for the project #${id}`}
```

bsekachev commented on cvat-ui/src/reducers/import-reducer.ts:
```diff
@@ -0,0 +1,67 @@
+import { ImportState } from './interfaces';
+
+const defaultState: ImportState = {
+    projects: {},
+    progress: 0.0,
+    status: '',
```
Several activities, but one progress and status looks strange

bsekachev replied:
Maybe instance is enough?

ActiveChooN replied:
Added format value to track active import proccess

bsekachev commented on cvat-ui/src/components/import-dataset-modal/import-dataset-modal.tsx:
```diff
@@ -0,0 +1,153 @@
+                visible={modalVisible}
+                onCancel={closeModal}
+                onOk={() => form.submit()}
+                className='cvat-modal-import-project'
```
destroyOnClose?

bsekachev replied:
```suggestion
                className='cvat-modal-import-dataset'
```

bsekachev commented on cvat-ui/src/components/import-dataset-modal/import-dataset-modal.tsx:
```diff
@@ -0,0 +1,153 @@
+    );
+}
+
+export default ImportDatasetModal;
```
React.memo()?

bsekachev commented on cvat-ui/src/components/import-dataset-modal/import-dataset-modal.tsx:
```diff
@@ -0,0 +1,153 @@
+            Notification.info({
+                message: 'Dataset export started',
+                description: `Dataset import was started for project #${instance?.id}. `,
+                className: 'cvat-notification-notice-import-project-start',
```
```suggestion
                className: 'cvat-notification-notice-import-dataset-start',
```

bsekachev commented on cvat-ui/src/components/import-dataset-modal/import-dataset-modal.tsx:
```diff
@@ -0,0 +1,153 @@
+                                                value={importer.name}
+                                                key={importer.name}
+                                                disabled={disabled}
+                                                className='cvat-modal-import-option-item'
```
```suggestion
                                                className='cvat-modal-import-dataset-option-item'
```

nmanovic commented on cvat/apps/dataset_manager/bindings.py:
```diff
@@ -3,27 +3,33 @@
 from .annotation import AnnotationIR, AnnotationManager, TrackManager
 
 
+CVAT_INTERNAL_ATTRIBUTES = {'occluded', 'outside', 'keyframe', 'track_id'}
```
rotation?

zhiltsov-max commented:
Recently added formats need to be adjusted as others.

bsekachev commented:
Once selected menu item keeps blue color when open menu again. Also need to align button content (Actions). Probably the button is to wide. 

![Screenshot from 2021-11-22 12-54-31](https://user-images.githubusercontent.com/40690378/142842534-65ebc408-9409-48f0-ade5-4d29fe3eb6ad.png)

Need to make icons blue (Like in export dialog). Need to add space between icons and text

![Screenshot from 2021-11-22 12-55-04](https://user-images.githubusercontent.com/40690378/142842806-98576147-b905-47d7-b6f8-da72d0666e2c.png)

Project #undefined
![Screenshot from 2021-11-22 13-03-20](https://user-images.githubusercontent.com/40690378/142842878-e6fdd589-8912-4cfe-bff6-7d5878bb2f2c.png)

One progress bar is enough. Progress component always shows 0.
![Screenshot from 2021-11-22 13-04-17](https://user-images.githubusercontent.com/40690378/142842924-ff414c85-5534-46cf-8b02-b12c242b76bc.png)


When import a dataset (dataset from a project with a couple of videos in train set, COCO format, can share the file if you need):
![Screenshot from 2021-11-22 13-12-22](https://user-images.githubusercontent.com/40690378/142843097-27e05e09-d7c9-48da-aeb8-0c2acc3693e3.png)


I got the same error as Maxim (Import job already exist) after failed import. 

I do not think it is related with this patch, but when I export a project (with one video in train) as CVAT for video, the archive had twice duplicated images
![Screenshot from 2021-11-22 13-20-44](https://user-images.githubusercontent.com/40690378/142844326-71afceef-626d-4b66-941e-a35229aa5be9.png)
![Screenshot from 2021-11-22 13-20-54](https://user-images.githubusercontent.com/40690378/142844328-457a6a37-5bd4-40c9-9526-f5bf11c6e8dc.png)
![Screenshot from 2021-11-22 13-21-01](https://user-images.githubusercontent.com/40690378/142844329-7e75d566-3c13-4445-9ab1-460f68ecff5c.png)
:

ActiveChooN commented:
@zhiltsov-max,

> 1. Recently added formats need to be adjusted as others.

Done.

> I see, we can only import dataset as a task into an existing project. Can we create a project from a dataset? It looks like it is supported on the server, but UI doesn't provide a button for this.

It was discussed in the planning stage and the solution was to keep only one place for dataset importing.
All problems with dataset importing were also solved.

@bsekachev 

> Once selected menu item keeps blue color when open menu again. Also need to align button content (Actions). Probably the button is to wide.

Done.

> Need to make icons blue (Like in export dialog). Need to add space between icons and text

Done.

> Project #undefined

Fixed.

> One progress bar is enough. Progress component always shows 0.

It was discussed. Probably will be enhanced in future PRs.

> When import a dataset (dataset from a project with a couple of videos in train set, COCO format, can share the file if you need):

Can you share the dataset, please?

> Empty alert while uploading .zip to the server. In the most of real cases it takes significant time to upload, so, better to write something to not to confuse a user.

Fixed.

> An attribute is_crowd appeared after trying to import (when importing COCO)

I guess it's normal behavior for the COCO dataset, @zhiltsov-max can correct me.

> Errors in console when open Actions menu:

Fixed.

bsekachev commented:
![image](https://user-images.githubusercontent.com/40690378/143565342-2ce1b8f2-27ba-4a5f-aa2a-1fa58e647ee2.png)
 Import error occure
[project_test-2021_11_24_13_47_06-cvat for images 1.1.zip](https://github.com/openvinotoolkit/cvat/files/7608048/project_test-2021_11_24_13_47_06-cvat.for.images.1.1.zip)
d

bsekachev commented:
>> Empty alert while uploading .zip to the server. In the most of real cases it takes significant time to upload, so, better to write something to not to confuse a user.

> Fixed.

I am not sure it was fixed. Now I see message: "The file is being uploaded to the server". Then it disappears and I see empty Alert about 15 seconds before Dataset importing started
![image](https://user-images.githubusercontent.com/40690378/143567970-4a3a12fd-f4f1-4150-a568-2c4e53185959.png)


>> One progress bar is enough. Progress component always shows 0.

> It was discussed. Probably will be enhanced in future PRs.

IMHO it is important UX issue and better to fix it ASAP.

bsekachev commented:
Some mess happens with annotations in imported task. See attached gif (the first is source task, the second is imported).
Extra objects appeared on frames 1,2,3, .. etc.
![source](https://user-images.githubusercontent.com/40690378/144563627-229731be-7815-4d18-9bc6-181cf0db2811.gif)


![result](https://user-images.githubusercontent.com/40690378/144563554-6f33181a-455e-4727-bb62-9d9e6adbda38.gif)

ActiveChooN commented:
@bsekachev 

> Mess happens with annotations

Fixed. Also fixed issue with second uploading.

bsekachev commented:
Could you please resolve conflict?

bsekachev commented:
Subset "default" was created when import a dataset with not specified subset, then I exported the project and tried to import it again:

![image](https://user-images.githubusercontent.com/40690378/145562536-61216d91-d5ce-41fa-b89e-e2fada393147.png)

Probably not related with this PR, but if I reopen dataset export modal before the previous export finished, I see infinite scroolling. If it is easy to fix, maybe fix it? If not, I can submit an issue. 
![image](https://user-images.githubusercontent.com/40690378/145562889-899b8353-b2bb-40fd-88cb-1b6f23bf1a4c.png)

bsekachev commented:
@ActiveChooN Please, take a look at tests

ActiveChooN commented:
> @ActiveChooN Please, take a look at tests

@bsekachev, fixed

bsekachev commented:
Could you update cvat-ui version?

ActiveChooN commented:
> Could you update cvat-ui version?

Sure

nmanovic commented:
@ActiveChooN , @bsekachev , I'm going to merge the patch today if I don't find any problems.

nmanovic reviewed (approved):
LGTM

ActiveChooN commented:
@aschernov, could you please update the documentation?
@dvkruchinin, could you please prepare UI tests?

dvkruchinin commented:
@ActiveChooN 
> could you please prepare UI tests?

Sure. A test will be prepared.
